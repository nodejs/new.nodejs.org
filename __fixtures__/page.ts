const toc1 = [
  { depth: 1, value: 'Heading 1', data: { id: 'custom-id' } },
  { depth: 2, value: 'Heading 2' },
  { depth: 3, value: 'Heading 3' },
];

const toc2 = [
  { depth: 1, value: 'Heading 1' },
  { depth: 3, value: 'Heading 3' },
  { depth: 3, value: 'Heading 3' },
  { depth: 2, value: 'Heading 2' },
  { depth: 4, value: 'Heading 4' },
  { depth: 6, value: 'Heading 6' },
  { depth: 5, value: 'Heading 5' },
  { depth: 1, value: 'Heading 1' },
];

export { toc1, toc2 };
