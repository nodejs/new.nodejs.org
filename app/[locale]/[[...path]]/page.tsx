import { notFound } from 'next/navigation';
import { unstable_setRequestLocale } from 'next-intl/server';
import type { FC } from 'react';

import { setClientContext } from '@/client-context';
import { MDXRenderer } from '@/components/mdxRenderer';
import { WithLayout } from '@/components/withLayout';
import { ENABLE_STATIC_EXPORT, VERCEL_REVALIDATE } from '@/next.constants.mjs';
import { DEFAULT_VIEWPORT } from '@/next.dynamic.constants.mjs';
import { dynamicRouter } from '@/next.dynamic.mjs';
import { availableLocaleCodes, defaultLocale } from '@/next.locales.mjs';
import { MatterProvider } from '@/providers/matterProvider';

type DynamicStaticPaths = { path: string[]; locale: string };
type DynamicParams = { params: DynamicStaticPaths };

// This is the default Viewport Metadata
// @see https://nextjs.org/docs/app/api-reference/functions/generate-viewport#generateviewport-function
export const generateViewport = async () => ({ ...DEFAULT_VIEWPORT });

// This generates each page's HTML Metadata
// @see https://nextjs.org/docs/app/api-reference/functions/generate-metadata
export const generateMetadata = async ({ params }: DynamicParams) => {
  const { path = [], locale = defaultLocale.code } = params;

  const pathname = dynamicRouter.getPathname(path);

  // Retrieves and rewriting rule if the pathname matches any rule
  const [, rewriteRule] = dynamicRouter.getRouteRewrite(pathname);

  return dynamicRouter.getPageMetadata(
    locale,
    rewriteRule ? rewriteRule(pathname) : pathname
  );
};

// This provides all the possible paths that can be generated statically
// + provides all the paths that we support on the Node.js Website
export const generateStaticParams = async () => {
  const paths: DynamicStaticPaths[] = [];

  // We don't need to compute all possible paths on regular builds
  // as we benefit from Next.js's ISR (Incremental Static Regeneration)
  if (!ENABLE_STATIC_EXPORT) {
    return [];
  }

  for (const locale of availableLocaleCodes) {
    const routesForLanguage = await dynamicRouter.getRoutesByLanguage(locale);

    const mappedRoutesWithLocale = routesForLanguage.map(pathname =>
      dynamicRouter.mapPathToRoute(locale, pathname)
    );

    paths.push(...mappedRoutesWithLocale);
  }

  return paths.sort();
};

// This method parses the current pathname and does any sort of modifications needed on the route
// then it proceeds to retrieve the Markdown file and parse the MDX Content into a React Component
// finally it returns (if the locale and route are valid) the React Component with the relevant context
// and attached context providers for rendering the current page
const getPage: FC<DynamicParams> = async ({ params }) => {
  const { path = [], locale = defaultLocale.code } = params;

  if (!availableLocaleCodes.includes(locale)) {
    // Forces the current locale to be the Default Locale
    unstable_setRequestLocale(defaultLocale.code);

    return notFound();
  }

  // Configures the current Locale to be the given Locale of the Request
  unstable_setRequestLocale(locale);

  const pathname = dynamicRouter.getPathname(path);

  if (dynamicRouter.shouldIgnoreRoute(pathname)) {
    return notFound();
  }

  // Retrieves and rewriting rule if the pathname matches any rule
  const [, rewriteRule] = dynamicRouter.getRouteRewrite(pathname);

  // We retrieve the source of the Markdown file by doing an educated guess
  // of what possible files could be the source of the page, since the extension
  // context is lost from `getStaticProps` as a limitation of Next.js itself
  const { source, filename } = await dynamicRouter.getMarkdownFile(
    locale,
    rewriteRule ? rewriteRule(pathname) : pathname
  );

  if (source.length && filename.length) {
    // This parses the source Markdown content and returns a React Component and
    // relevant context from the Markdown File
    const { MDXContent, frontmatter, headings, readingTime } =
      await dynamicRouter.getMDXContent(source, filename);

    // Metadata and shared Context to be available through the lifecycle of the page
    const sharedContext = {
      frontmatter,
      headings,
      pathname: `/${pathname}`,
      readingTime,
      filename,
    };

    // Defines a shared Server Context for the Client-Side
    // That is shared for all pages under the dynamic router
    setClientContext(sharedContext);

    // The Matter Provider allows Client-Side injection of the data
    // to a shared React Client Provider even though the page is rendered
    // within a server-side context
    return (
      <MatterProvider {...sharedContext}>
        <WithLayout layout={frontmatter.layout}>
          <MDXRenderer Component={MDXContent} />
        </WithLayout>
      </MatterProvider>
    );
  }

  return notFound();
};

// In this case we want to catch-all possible pages even to this page. This ensures that we use our 404
// and that all pages including existing ones are handled here and provide `next-intl` locale also
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamicparams
export const dynamicParams = true;

// Enforces that this route is used as static rendering
// Except whenever on the Development mode as we want instant-refresh when making changes
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic
export const dynamic = 'error';

// Ensures that this endpoint is invalidated and re-executed every X minutes
// so that when new deployments happen, the data is refreshed
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate
export const revalidate = VERCEL_REVALIDATE;

export default getPage;
