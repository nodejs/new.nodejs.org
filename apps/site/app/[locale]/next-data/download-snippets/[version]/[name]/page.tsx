import { notFound } from 'next/navigation';
import type { FC } from 'react';
import { VFile } from 'vfile';

import JSXCodeBox from '@/components/JSX/CodeBox';
import provideDownloadSnippets from '@/next-data/providers/downloadSnippets';
import provideReleaseData from '@/next-data/providers/releaseData';
import { defaultLocale } from '@/next.locales.mjs';
import { compile } from '@/next.mdx.compiler.mjs';

type DynamicStaticPaths = { locale: string; version: string; name: string };
type StaticParams = { params: Promise<DynamicStaticPaths> };

// This is the Route Handler for the `GET` method which handles the request
// for generating OpenGraph images for Blog Posts and Pages
// @see https://nextjs.org/docs/app/building-your-application/routing/router-handlers
const getPage: FC<StaticParams> = async props => {
  const params = await props.params;

  const releaseData = provideReleaseData();
  const snippets = provideDownloadSnippets(params.locale);

  const release = releaseData.find(({ version }) => version === params.version);
  const snippet = snippets?.find(({ name }) => name === params.name);

  // If the release data, snippets data, and the snippet are available we proceed
  // Since we need to find a matching Node.js release metadata, and also render the code snippet
  // for the correct locale, version and snippet name (i.e.: nvm)
  if (snippets && release && snippet) {
    // Wraps the code snippet (which is a plain string imported from `provideDownloadSnippets`)
    // To be wrapped with the JSX CodeBox component that invokes Shiki's Hast to JSX runtime compiler
    // Whilst evaluating the snippet as a JavaScript template literal, which allows us to access
    // the `props` object passed to the MDX Compiler.
    const wrappedSource = `<CodeBox language="${snippet.language}">{\`${snippet.content}\`}</CodeBox>`;

    // Compiles the wrapped code snippet with the MDX Compiler
    const { content } = await compile(
      // Wraps the code snippet with a VFile object to be passed to the MDX Compiler
      new VFile(wrappedSource),
      // The MDX Compiler will use the `jsx` extension to compile the code snippet
      'mdx',
      // Passes the `CodeBox` component to the MDX Compiler as a component within the `components` object
      { CodeBox: JSXCodeBox },
      // Passes the `release` object to the MDX Compiler as a property within the `props` object
      { release }
    );

    return content;
  }

  return notFound();
};

// This function generates the static paths that come from the dynamic segments
// `[locale]/next-data/download-snippets/[version]/[name]` this will return a default value as we don't want to
// statically generate this route as it is compute-expensive.
// Hence we generate a fake route just to satisfy Next.js requirements.
export const generateStaticParams = async () => [
  {
    locale: defaultLocale.code,
    version: 'latest',
    name: 'default',
  },
];

// Enforces that this route is cached and static as much as possible
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic
export const dynamic = 'force-static';

export default getPage;
