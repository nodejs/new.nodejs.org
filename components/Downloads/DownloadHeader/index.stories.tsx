import DownloadHeader from '.';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof DownloadHeader>;
type Meta = MetaObj<typeof DownloadHeader>;

export const Default: Story = {};

export default { component: DownloadHeader } as Meta;
