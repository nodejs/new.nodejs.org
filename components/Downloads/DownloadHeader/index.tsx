import { FormattedMessage, injectIntl } from 'react-intl';
import styles from './index.module.scss';
import SectionTitle from '../../Common/SectionTitle';
import type { WrappedComponentProps } from 'react-intl';
import type { NodeReleaseData } from '../../../types';

type DownloadHeaderProps = { release: NodeReleaseData }
  release?: NodeReleaseData;
};

const DownloadHeader = ({ release, intl }: Props & WrappedComponentProps) => (
  <>
    <div className={styles.downloadHeader}>
      <SectionTitle
        path={[
          'home',
          intl.formatMessage({
            id: 'components.downloadHeader.navigation.activeSection',
          }),
        ]}
      />
      <div>
        <FormattedMessage
          id="components.downloadHeader.navigation.nodeVersion"
          values={{ lts: release?.isLts, nodeVersion: release?.fullVersion }}
        />
      </div>
    </div>
    <div className={styles.downloadHeader}>
      <div className={styles.title}>
        <FormattedMessage id="components.downloadHeader.navigation.title" />
      </div>
      <div className={styles.npm}>
        <FormattedMessage id="components.downloadHeader.navigation.npmVersion" />
      </div>
    </div>
  </>
);

export default injectIntl(DownloadHeader);
