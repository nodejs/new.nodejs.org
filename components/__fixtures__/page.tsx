export const mockTableOfContents = {
  items: [
    { title: 'mock-title', url: '#mock-title' },
    { title: 'mock-title-2', url: '#mock-title-2' },
  ],
};
