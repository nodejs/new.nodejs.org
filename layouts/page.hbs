<!DOCTYPE html>
<html lang="{{site.locale}}">
{{> html-head }}

<body>
  {{> header }}

  <div id="main">
    <div class="container">
      {{{ contents }}}
    </div>
  </div>

  {{> footer }}
</body>
</html>
