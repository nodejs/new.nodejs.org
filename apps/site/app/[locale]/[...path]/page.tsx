/**
 * This file extends on the `page.tsx` file, which is the default file that is used to render
 * the entry points for each locale and then also reused within the [...path] route to render the
 * and contains all logic for rendering our dynamic and static routes within the Node.js Website.
 *
 * Note: that each `page.tsx` should have its own `generateStaticParams` to prevent clash of
 * dynamic params, which will lead on static export errors and other sort of issues.
 */

import * as basePage from '@/app/[locale]/page';
import { ENABLE_STATIC_EXPORT } from '@/next.constants.mjs';
import { dynamicRouter } from '@/next.dynamic.mjs';
import { availableLocaleCodes } from '@/next.locales.mjs';

// This is the default Viewport Metadata
// @see https://nextjs.org/docs/app/api-reference/functions/generate-viewport#generateviewport-function
export const generateViewport = basePage.generateViewport;

// This generates each page's HTML Metadata
// @see https://nextjs.org/docs/app/api-reference/functions/generate-metadata
export const generateMetadata = basePage.generateMetadata;

// This provides all the possible paths that can be generated statically
// + provides all the paths that we support on the Node.js Website
export const generateStaticParams = async () => {
  const allAvailableRoutes = await Promise.all(
    // Gets all mapped routes to the Next.js Routing Engine by Locale
    availableLocaleCodes.map(async (locale: string) => {
      const routesForLanguage = await dynamicRouter.getRoutesByLanguage(locale);

      return routesForLanguage.map(pathname =>
        dynamicRouter.mapPathToRoute(locale, pathname)
      );
    })
  );

  return ENABLE_STATIC_EXPORT ? allAvailableRoutes.flat().sort() : [];
};

// Enforces that this route is used as static rendering
// Except whenever on the Development mode as we want instant-refresh when making changes
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#dynamic
export const dynamic = 'force-static';

// Ensures that this endpoint is invalidated and re-executed every X minutes
// so that when new deployments happen, the data is refreshed
// @see https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config#revalidate
export const revalidate = 300;

export default basePage.default;
