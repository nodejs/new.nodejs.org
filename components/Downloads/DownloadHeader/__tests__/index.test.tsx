import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import DownloadHeader from '../';
import { createNodeReleasesData } from '../../../../__fixtures__/page';

describe('DownloadHeader component', (): void => {
  it('renders correctly', (): void => {
    const { container } = render(
      <IntlProvider locale="en" onError={() => {}}>
        <DownloadHeader release={createNodeReleasesData()[0]} />
      </IntlProvider>
    );
    expect(container).toMatchSnapshot();
  });
});
