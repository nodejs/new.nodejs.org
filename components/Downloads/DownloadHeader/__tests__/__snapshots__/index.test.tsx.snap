// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadHeader component renders correctly 1`] = `
<div>
  <div
    class="downloadHeader"
  >
    <div
      class="sectionTitle"
    >
      home / 
      <span
        class="active"
      >
        components.downloadHeader.navigation.activeSection
      </span>
    </div>
    <div>
      components.downloadHeader.navigation.nodeVersion
    </div>
  </div>
  <div
    class="downloadHeader"
  >
    <div
      class="title"
    >
      components.downloadHeader.navigation.title
    </div>
    <div
      class="npm"
    >
      components.downloadHeader.navigation.npmVersion
    </div>
  </div>
</div>
`;
